File: trunk/dotnet/picocontainer/src/Test/Defaults/ChildContainerTestCase.cs
using PicoContainer;
using PicoContainer.Defaults;
using NUnit.Framework;
using PicoContainer.Tests.TestModel;

namespace Test.Defaults
{
	[TestFixture]
	public class ChildContainerTestCase
	{
		public void testParentContainerWithComponentWithEqualKeyShouldBeShadowedByChild()
		{
			DefaultPicoContainer parent = new DefaultPicoContainer();
			DefaultPicoContainer child = new DefaultPicoContainer(parent);

			parent.RegisterComponentImplementation("key", typeof (AlternativeTouchable));
			child.RegisterComponentImplementation("key", typeof (SimpleTouchable));
			child.RegisterComponentImplementation(typeof (DependsOnTouchable));

			DependsOnTouchable dot = (DependsOnTouchable) child.GetComponentInstanceOfType(typeof (DependsOnTouchable));
			Assert.AreEqual(typeof (SimpleTouchable), dot.getTouchable().GetType());
		}

		public void testParentComponentRegisteredAsClassShouldBePreffered()
		{
			DefaultPicoContainer parent = new DefaultPicoContainer();
			DefaultPicoContainer child = new DefaultPicoContainer(parent);

			parent.RegisterComponentImplementation(typeof (Touchable), typeof (AlternativeTouchable));
			child.RegisterComponentImplementation("key", typeof (SimpleTouchable));
			child.RegisterComponentImplementation(typeof (DependsOnTouchable));

			DependsOnTouchable dot = (DependsOnTouchable) child.GetComponentInstanceOfType(typeof (DependsOnTouchable));
			Assert.AreEqual(typeof (AlternativeTouchable), dot.getTouchable().GetType());
		}

		public void testResolveFromParentByType()
		{
			IMutablePicoContainer parent = new DefaultPicoContainer();
			parent.RegisterComponentImplementation(typeof (Touchable), typeof (SimpleTouchable));

			IMutablePicoContainer child = new DefaultPicoContainer(parent);
			child.RegisterComponentImplementation(typeof (DependsOnTouchable));

			Assert.IsNotNull(child.GetComponentInstance(typeof (DependsOnTouchable)));
		}

		public void testResolveFromParentByKey()
		{
			IMutablePicoContainer parent = new DefaultPicoContainer();
			parent.RegisterComponentImplementation(typeof (Touchable), typeof (SimpleTouchable));

			IMutablePicoContainer child = new DefaultPicoContainer(parent);
			child.RegisterComponentImplementation(typeof (DependsOnTouchable), typeof (DependsOnTouchable),
			                                      new IParameter[] {new ComponentParameter(typeof (Touchable))});

			Assert.IsNotNull(child.GetComponentInstance(typeof (DependsOnTouchable)));
		}

		public void testResolveFromGrandParentByType()
		{
			IMutablePicoContainer grandParent = new DefaultPicoContainer();
			grandParent.RegisterComponentImplementation(typeof (Touchable), typeof (SimpleTouchable));

			IMutablePicoContainer parent = new DefaultPicoContainer(grandParent);

			IMutablePicoContainer child = new DefaultPicoContainer(parent);
			child.RegisterComponentImplementation(typeof (DependsOnTouchable));

			Assert.IsNotNull(child.GetComponentInstance(typeof (DependsOnTouchable)));
		}

		public void testResolveFromGrandParentByKey()
		{
			IMutablePicoContainer grandParent = new DefaultPicoContainer();
			grandParent.RegisterComponentImplementation(typeof (Touchable), typeof (SimpleTouchable));

			IMutablePicoContainer parent = new DefaultPicoContainer(grandParent);

			IMutablePicoContainer child = new DefaultPicoContainer(parent);
			child.RegisterComponentImplementation(typeof (DependsOnTouchable), typeof (DependsOnTouchable),
			                                      new IParameter[] {new ComponentParameter(typeof (Touchable))});

			Assert.IsNotNull(child.GetComponentInstance(typeof (DependsOnTouchable)));
		}

	}
}

File: java/1.x/picocontainer/tags/picocontainer-1_2-beta-1/container/src/test/org/picocontainer/defaults/ChildContainerTestCase.java
/*****************************************************************************
 * Copyright (C) PicoContainer Organization. All rights reserved.            *
 * ------------------------------------------------------------------------- *
 * The software in this package is published under the terms of the BSD      *
 * style license a copy of which has been included with this distribution in *
 * the LICENSE.txt file.                                                     *
 *                                                                           *
 * Original code by                                                          *
 *****************************************************************************/
package org.picocontainer.defaults;

import junit.framework.TestCase;
import org.picocontainer.MutablePicoContainer;
import org.picocontainer.Parameter;
import org.picocontainer.testmodel.AlternativeTouchable;
import org.picocontainer.testmodel.DependsOnTouchable;
import org.picocontainer.testmodel.SimpleTouchable;
import org.picocontainer.testmodel.Touchable;

/**
 * @author Paul Hammant
 * @author Aslak Helles&oslash;y
 * @author Rafal Krzewski
 * @version $Revision$
 */
public class ChildContainerTestCase extends TestCase {

    public void testParentContainerWithComponentWithEqualKeyShouldBeShadowedByChild() throws Exception {
        DefaultPicoContainer parent = new DefaultPicoContainer();
        DefaultPicoContainer child = new DefaultPicoContainer(parent);

        parent.registerComponentImplementation("key", AlternativeTouchable.class);
        child.registerComponentImplementation("key", SimpleTouchable.class);
        child.registerComponentImplementation(DependsOnTouchable.class);

        DependsOnTouchable dot = (DependsOnTouchable) child.getComponentInstanceOfType(DependsOnTouchable.class);
        assertEquals(SimpleTouchable.class, dot.getTouchable().getClass());
    }

    public void testParentComponentRegisteredAsClassShouldBePreffered() throws Exception {
        DefaultPicoContainer parent = new DefaultPicoContainer();
        DefaultPicoContainer child = new DefaultPicoContainer(parent);

        parent.registerComponentImplementation(Touchable.class, AlternativeTouchable.class);
        child.registerComponentImplementation("key", SimpleTouchable.class);
        child.registerComponentImplementation(DependsOnTouchable.class);

        DependsOnTouchable dot = (DependsOnTouchable) child.getComponentInstanceOfType(DependsOnTouchable.class);
        assertEquals(AlternativeTouchable.class, dot.getTouchable().getClass());
    }

    public void testResolveFromParentByType() {
        MutablePicoContainer parent = new DefaultPicoContainer();
        parent.registerComponentImplementation(Touchable.class, SimpleTouchable.class);

        MutablePicoContainer child = new DefaultPicoContainer(parent);
        child.registerComponentImplementation(DependsOnTouchable.class);

        assertNotNull(child.getComponentInstance(DependsOnTouchable.class));
    }

    public void testResolveFromParentByKey() {
        MutablePicoContainer parent = new DefaultPicoContainer();
        parent.registerComponentImplementation(Touchable.class, SimpleTouchable.class);

        MutablePicoContainer child = new DefaultPicoContainer(parent);
        child.registerComponentImplementation(DependsOnTouchable.class, DependsOnTouchable.class,
                new Parameter[]{new ComponentParameter((Object) Touchable.class)});

        assertNotNull(child.getComponentInstance(DependsOnTouchable.class));
    }

    public void testResolveFromGrandParentByType() {
        MutablePicoContainer grandParent = new DefaultPicoContainer();
        grandParent.registerComponentImplementation(Touchable.class, SimpleTouchable.class);

        MutablePicoContainer parent = new DefaultPicoContainer(grandParent);

        MutablePicoContainer child = new DefaultPicoContainer(parent);
        child.registerComponentImplementation(DependsOnTouchable.class);

        assertNotNull(child.getComponentInstance(DependsOnTouchable.class));
    }

    public void testResolveFromGrandParentByKey() {
        MutablePicoContainer grandParent = new DefaultPicoContainer();
        grandParent.registerComponentImplementation(Touchable.class, SimpleTouchable.class);

        MutablePicoContainer parent = new DefaultPicoContainer(grandParent);

        MutablePicoContainer child = new DefaultPicoContainer(parent);
        child.registerComponentImplementation(DependsOnTouchable.class, DependsOnTouchable.class,
                new Parameter[]{new ComponentParameter((Object) Touchable.class)});

        assertNotNull(child.getComponentInstance(DependsOnTouchable.class));
    }
}
